File: tyhp-lang/phpstorm/src/unitTest/kotlin/com/tyhp/lang/debug/ProxyProcessControllerTest.kt
package com.tyhp.lang.debug

import java.util.concurrent.CopyOnWriteArrayList
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertFalse
import kotlin.test.assertTrue

class ProxyProcessControllerTest {
    private fun createHarness(
        listenPort: Int? = null,
        failListen: String? = null,
    ): Harness {
        val spawned = CopyOnWriteArrayList<FakeSpawnedProxy>()
        val states = CopyOnWriteArrayList<ProxyRunState>()
        val controller = ProxyProcessController(
            object : ProxyProcessHooks {
                override fun spawn(command: String, args: List<String>, cwd: String?): SpawnedProxy {
                    val child = FakeSpawnedProxy()
                    spawned.add(child)
                    return child
                }

                override fun waitForListening(port: Int, child: SpawnedProxy, abort: AtomicBoolean): Int {
                    if (abort.get()) {
                        throw IllegalStateException("aborted")
                    }
                    if (failListen != null) {
                        throw IllegalStateException(failListen)
                    }
                    return listenPort ?: port
                }

                override fun stopProcess(child: SpawnedProxy) {
                    child.destroyTerm()
                }

                override fun onState(state: ProxyRunState) {
                    states.add(state)
                }
            },
        )
        return Harness(controller, spawned, states)
    }

    @Test
    fun `start then listening reaches running`() {
        val harness = createHarness()
        harness.controller.start(
            ProxyStartRequest(
                command = "/bin/tyhp",
                args = listOf("xdebug_proxy", "--ide-port=9003", "--xdebug-port=9004"),
                idePort = 9003,
            ),
        ).get(5, TimeUnit.SECONDS)
        assertEquals(ProxyRunState.RUNNING, harness.controller.currentState)
        assertEquals(1, harness.spawned.size)
        assertFalse(harness.spawned[0].killed)
        harness.controller.shutdownExecutor()
    }

    @Test
    fun `start while running is a no-op and does not spawn a second process`() {
        val harness = createHarness()
        val request = ProxyStartRequest(command = "/bin/tyhp", args = listOf("xdebug_proxy"), idePort = 9003)
        harness.controller.start(request).get(5, TimeUnit.SECONDS)
        harness.controller.start(request).get(5, TimeUnit.SECONDS)
        assertEquals(1, harness.spawned.size)
        assertEquals(ProxyRunState.RUNNING, harness.controller.currentState)
        harness.controller.shutdownExecutor()
    }

    @Test
    fun `stop signals SIGTERM via the manager stop path and reaches stopped`() {
        val harness = createHarness()
        harness.controller.start(
            ProxyStartRequest(command = "/bin/tyhp", args = listOf("xdebug_proxy"), idePort = 9003),
        ).get(5, TimeUnit.SECONDS)
        harness.controller.stop().get(5, TimeUnit.SECONDS)
        assertEquals(ProxyRunState.STOPPED, harness.controller.currentState)
        assertTrue(harness.spawned[0].killed)
        assertEquals("SIGTERM", harness.spawned[0].lastSignal)
        harness.controller.shutdownExecutor()
    }

    @Test
    fun `restart stops then starts a new process`() {
        val harness = createHarness()
        val request = ProxyStartRequest(command = "/bin/tyhp", args = listOf("xdebug_proxy"), idePort = 9003)
        harness.controller.start(request).get(5, TimeUnit.SECONDS)
        harness.controller.restart(request).get(5, TimeUnit.SECONDS)
        assertEquals(ProxyRunState.RUNNING, harness.controller.currentState)
        assertEquals(2, harness.spawned.size)
        assertTrue(harness.spawned[0].killed)
        assertFalse(harness.spawned[1].killed)
        harness.controller.shutdownExecutor()
    }

    @Test
    fun `listen failure leaves error state and stops the child`() {
        val harness = createHarness(failListen = "port in use")
        val error = assertFailsWith<Exception> {
            harness.controller.start(
                ProxyStartRequest(command = "/bin/tyhp", args = listOf("xdebug_proxy"), idePort = 9003),
            ).get(5, TimeUnit.SECONDS)
        }
        assertTrue(error.message?.contains("port in use") == true || error.cause?.message?.contains("port in use") == true)
        assertEquals(ProxyRunState.ERROR, harness.controller.currentState)
        assertTrue(harness.spawned[0].killed)
        harness.controller.shutdownExecutor()
    }

    @Test
    fun `unexpected exit from running becomes error`() {
        val harness = createHarness()
        harness.controller.start(
            ProxyStartRequest(command = "/bin/tyhp", args = listOf("xdebug_proxy"), idePort = 9003),
        ).get(5, TimeUnit.SECONDS)
        harness.spawned[0].exit(1)
        assertEquals(ProxyRunState.ERROR, harness.controller.currentState)
        harness.controller.shutdownExecutor()
    }

    @Test
    fun `stopSpawnedProxy uses SIGTERM and waits for exit`() {
        val child = FakeSpawnedProxy()
        stopSpawnedProxy(child)
        assertTrue(child.killed)
        assertEquals("SIGTERM", child.lastSignal)
        assertEquals(0, child.exitCode)
    }

    private data class Harness(
        val controller: ProxyProcessController,
        val spawned: List<FakeSpawnedProxy>,
        val states: List<ProxyRunState>,
    )
}

File: tyhp-lang/vscode/src/debug/ProxyProcessController.test.ts
import assert from "node:assert/strict";
import { test } from "node:test";
import {
    FakeSpawnedProxy,
    ProxyProcessController,
    stopSpawnedProxy,
} from "./ProxyProcessController";

function createHarness(options?: { listenPort?: number; failListen?: string }) {
    const spawned: FakeSpawnedProxy[] = [];
    const states: string[] = [];
    const controller = new ProxyProcessController({
        spawn: () => {
            const child = new FakeSpawnedProxy();
            spawned.push(child);
            return child;
        },
        waitForListening: async (port, _child, abort) => {
            if (abort.aborted) {
                throw new Error("aborted");
            }
            if (options?.failListen) {
                throw new Error(options.failListen);
            }
            return options?.listenPort ?? port;
        },
        stopProcess: async (child) => {
            child.kill("SIGTERM");
        },
        onState: (state) => states.push(state),
    });
    return { controller, spawned, states };
}

test("start then listening reaches running", async () => {
    const { controller, spawned } = createHarness();
    await controller.start({
        command: "/bin/tyhp",
        args: ["xdebug_proxy", "--ide-port=9003", "--xdebug-port=9004"],
        idePort: 9003,
    });
    assert.equal(controller.currentState, "running");
    assert.equal(spawned.length, 1);
    assert.equal(spawned[0]?.killed, false);
});

test("start while running is a no-op (does not spawn a second process)", async () => {
    const { controller, spawned } = createHarness();
    const request = { command: "/bin/tyhp", args: ["xdebug_proxy"], idePort: 9003 };
    await controller.start(request);
    await controller.start(request);
    assert.equal(spawned.length, 1);
    assert.equal(controller.currentState, "running");
});

test("stop signals SIGTERM via the manager stop path and reaches stopped", async () => {
    const { controller, spawned } = createHarness();
    await controller.start({ command: "/bin/tyhp", args: ["xdebug_proxy"], idePort: 9003 });
    await controller.stop();
    assert.equal(controller.currentState, "stopped");
    assert.equal(spawned[0]?.killed, true);
    assert.equal(spawned[0]?.lastSignal, "SIGTERM");
});

test("restart stops then starts a new process", async () => {
    const { controller, spawned } = createHarness();
    const request = { command: "/bin/tyhp", args: ["xdebug_proxy"], idePort: 9003 };
    await controller.start(request);
    await controller.restart(request);
    assert.equal(controller.currentState, "running");
    assert.equal(spawned.length, 2);
    assert.equal(spawned[0]?.killed, true);
    assert.equal(spawned[1]?.killed, false);
});

test("listen failure leaves error state and stops the child", async () => {
    const { controller, spawned } = createHarness({ failListen: "port in use" });
    await assert.rejects(
        () => controller.start({ command: "/bin/tyhp", args: ["xdebug_proxy"], idePort: 9003 }),
        /port in use/
    );
    assert.equal(controller.currentState, "error");
    assert.equal(spawned[0]?.killed, true);
});

test("unexpected exit from running becomes error", async () => {
    const { controller, spawned } = createHarness();
    await controller.start({ command: "/bin/tyhp", args: ["xdebug_proxy"], idePort: 9003 });
    spawned[0]?.exit(1);
    assert.equal(controller.currentState, "error");
});

test("stopSpawnedProxy uses SIGTERM and waits for exit", async () => {
    const child = new FakeSpawnedProxy();
    await stopSpawnedProxy(child);
    assert.equal(child.killed, true);
    assert.equal(child.lastSignal, "SIGTERM");
    assert.equal(child.exitCode, 0);
});
